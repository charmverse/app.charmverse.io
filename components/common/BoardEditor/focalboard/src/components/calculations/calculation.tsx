import ExpandLessIcon from '@mui/icons-material/ExpandLess';
import type { CSSProperties } from 'react';
import React from 'react';
import { useIntl } from 'react-intl';

import type { IPropertyTemplate } from '../../blocks/board';
import type { Card } from '../../blocks/card';

import Calculations from './calculations';
import type { CommonCalculationOptionProps } from './options';
import { Options, optionDisplayNameString } from './options';

type Props = {
  style: CSSProperties;
  class: string;
  value: string;
  menuOpen: boolean;
  onMenuClose: () => void;
  onMenuOpen: () => void;
  onChange: (value: string) => void;
  cards: readonly Card[];
  property: IPropertyTemplate;
  hovered: boolean;
  // eslint-disable-next-line react/no-unused-prop-types
  optionsComponent: React.ComponentType<CommonCalculationOptionProps>;
}

function Calculation (props: Props): JSX.Element {
  const value = props.value || Options.none.value;
  const valueOption = Options[value];
  const intl = useIntl();

  const option = (
    <props.optionsComponent
      value={value}
      menuOpen={props.menuOpen}
      onClose={props.onMenuClose}
      onChange={props.onChange}
    />
  );

  return (

  // tabindex is needed to make onBlur work on div.
  // See this for more details-
  // https://stackoverflow.com/questions/47308081/onblur-event-is-not-firing
    <div
      className={`Calculation ${value} ${props.class} ${props.menuOpen ? 'menuOpen' : ''} ${props.hovered ? 'hovered' : ''}`}
      style={props.style}
      onClick={() => (props.menuOpen ? props.onMenuClose() : props.onMenuOpen())}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex={0}
      onBlur={props.onMenuClose}
    >
      {
                props.menuOpen && (
                  <div>
                    {option}
                  </div>
                )
            }

      <span className='calculationLabel'>
        {optionDisplayNameString(valueOption!, intl)}
      </span>

      {
                value === Options.none.value
                && <ExpandLessIcon fontSize='small' />
            }

      {
                value !== Options.none.value
                && (
                  <span className='calculationValue'>
                    {Calculations[value] ? Calculations[value](props.cards, props.property, intl) : ''}
                  </span>
                )
            }

    </div>
  );
}

export default Calculation;
