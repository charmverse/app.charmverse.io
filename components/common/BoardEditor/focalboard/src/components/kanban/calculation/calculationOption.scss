.KanbanCalculationOptions_CustomOption {
  color: rgba(var(--center-channel-color-rgb), 1);
  min-height: 24px;
  max-width: 220px;
  width: 200px;
  padding: 2px 12px;
  cursor: pointer;
  position: relative;

  &:hover {
    background: var(--charmeditor-active-hover);
  }

  span {
    overflow: hidden;
  }

  .drops {
    padding: 2px 14px;
    width: 200px;

    &.active,
    &.active:hover {
      background-color: var(--charmeditor-active);
      color: rgb(var(--sidebar-text-rgb));
    }
  }

  .customs:hover,
  .drops:hover {
    background: var(--charmeditor-active-hover);
  }

  .dropdown-submenu {
    position: fixed;
    overflow: auto;
    background: var(--background-paper);
    border: 0;
    box-shadow: 0 0 0 1px hsla(0, 0%, 0%, 0.1), 0 4px 11px hsla(0, 0%, 0%, 0.1);
    z-index: 999;
    padding: 6px 0;
    border-radius: 4px;
  }

  .CompassIcon.icon-chevron-right {
    float: right;
  }

  &.active {
    background-color: var(--charmeditor-active);

    > span {
      color: rgb(var(--sidebar-text-rgb));
    }
  }
}

.CalculationOptions__menu {
  .CalculationOptions__menu-list {
    max-height: 400px;
    padding: 8px 0;
  }
}