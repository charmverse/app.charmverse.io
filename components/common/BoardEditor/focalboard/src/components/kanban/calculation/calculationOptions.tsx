
import type { IPropertyTemplate, PropertyType } from '../../../blocks/board';
import type {
  CommonCalculationOptionProps } from '../../calculations/options';
import {
  CalculationOptions,
  optionsByType
} from '../../calculations/options';

import type { OptionProps } from './kanbanOption';
import { Option } from './kanbanOption';

type Props = CommonCalculationOptionProps & {
    cardProperties: IPropertyTemplate[];
    onChange: (data: { calculation: string, propertyId: string }) => void;
    property: IPropertyTemplate;
}

// contains mapping of property types which are effectly the same as other property type.
const equivalentPropertyType = new Map<PropertyType, PropertyType>([
  ['createdTime', 'date'],
  ['updatedTime', 'date']
]);

export function getEquivalentPropertyType (propertyType: PropertyType): PropertyType {
  return equivalentPropertyType.get(propertyType) || propertyType;
}

export function KanbanCalculationOptions (props: Props): JSX.Element {
  const options: OptionProps[] = [];

    // Show common options, first,
    // followed by type-specific functions
    optionsByType.get('common')!.forEach((typeOption) => {
      if (typeOption.value !== 'none') {
        options.push({
          ...typeOption,
          cardProperties: props.cardProperties,
          onChange: props.onChange,
          activeValue: props.value,
          activeProperty: props.property
        });
      }
    });

    const seen: Record<string, boolean> = {};
    props.cardProperties.forEach((property) => {
      // skip already processed property types
      if (seen[getEquivalentPropertyType(property.type)]) {
        return;
      }

      (optionsByType.get(property.type) || [])
        .forEach((typeOption) => {
          options.push({
            ...typeOption,
            cardProperties: props.cardProperties,
            onChange: props.onChange,
            activeValue: props.value,
            activeProperty: props.property!
          });
        });

      seen[getEquivalentPropertyType(property.type)] = true;
    });

    return (
      <CalculationOptions
        value={props.value}
        menuOpen={props.menuOpen}
        onClose={props.onClose}
        onChange={props.onChange}
        options={options}
        components={{ Option }}
      />
    );
}
