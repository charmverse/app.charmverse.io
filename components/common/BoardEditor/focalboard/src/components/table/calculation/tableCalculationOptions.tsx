import React from 'react';

import type { CommonCalculationOptionProps } from '../../calculations/options';
import { CalculationOptions, optionsByType } from '../../calculations/options';

export function TableCalculationOptions (props: CommonCalculationOptionProps & { property?: { type: string } }): JSX.Element {
  const options = [...optionsByType.get('common')!];
  if (props.property && optionsByType.get(props.property.type)) {
    options.push(...optionsByType.get(props.property.type)!);
  }

  return (
    <CalculationOptions
      value={props.value}
      menuOpen={props.menuOpen}
      onClose={props.onClose}
      onChange={props.onChange}
      options={options}
    />
  );
}
