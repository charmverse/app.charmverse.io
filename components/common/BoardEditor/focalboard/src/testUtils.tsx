import type { PageMeta } from '@charmverse/core/pages';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { IntlProvider } from 'react-intl';
import type { Middleware } from 'redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureStore from 'redux-mock-store';
import { v4 } from 'uuid';

import { PagesContext } from 'hooks/usePages';
import type { Block } from 'lib/focalboard/block';
import { pageStubToCreate } from 'testing/generatePageStub';
import { AppThemeProvider } from 'theme/AppThemeProvider';

export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
  <AppThemeProvider>
    <IntlProvider locale='en'>{children}</IntlProvider>
  </AppThemeProvider>
);
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
  return <DndProvider backend={HTML5Backend}>{wrapIntl(children)}</DndProvider>;
};

export const wrapPagesProvider = (cardId: string | string[], children: React.ReactNode) => {
  const pages: Record<string, PageMeta> = {};

  const cardIds = Array.isArray(cardId) ? cardId : [cardId];

  cardIds.forEach((_cardId) => {
    pages[_cardId] = pageStubToCreate({ spaceId: v4(), createdBy: v4(), id: _cardId, path: _cardId }) as PageMeta;
  });

  return (
    <PagesContext.Provider
      value={
        {
          pages
        } as any
      }
    >
      {children}
    </PagesContext.Provider>
  );
};

export function mockDOM(): void {
  window.focus = jest.fn();
  document.createRange = () => {
    const range = new Range();
    range.getBoundingClientRect = jest.fn();
    range.getClientRects = () => {
      return {
        item: () => null,
        length: 0,
        [Symbol.iterator]: jest.fn()
      };
    };
    return range;
  };
}
export function mockMatchMedia(result: { matches: boolean }): void {
  // We check if system preference is dark or light theme.
  // This is required to provide it's definition since
  // window.matchMedia doesn't exist in Jest.
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => {
      return result;

      // return ({
      //     matches: true,
      // })
    })
  });
}

export function mockStateStore(middleware: Middleware[], state: unknown): MockStoreEnhanced<unknown, unknown> {
  const mockStore = configureStore(middleware);
  return mockStore(state);
}

export type BlocksById<BlockType> = { [key: string]: BlockType };

export function blocksById<BlockType extends Block>(blocks: BlockType[]): BlocksById<BlockType> {
  return blocks.reduce((res, block) => {
    res[block.id] = block;
    return res;
  }, {} as BlocksById<BlockType>);
}
