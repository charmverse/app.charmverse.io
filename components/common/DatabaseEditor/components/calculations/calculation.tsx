import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import type { CSSProperties } from 'react';
import React from 'react';
import { useIntl } from 'react-intl';

import type { IPropertyTemplate } from 'lib/databases/board';
import type { Card } from 'lib/databases/card';

import Calculations from './calculations';
import type { CommonCalculationOptionProps } from './options';
import { Options, optionDisplayNameString } from './options';

type Props = {
  style: CSSProperties;
  class: string;
  value: string;
  menuOpen: boolean;
  onMenuClose: () => void;
  onMenuOpen: (element: HTMLElement) => void;
  onChange: (value: string) => void;
  cards: readonly Card[];
  property: IPropertyTemplate;
  hovered: boolean;
  // eslint-disable-next-line react/no-unused-prop-types
  optionsComponent: React.ComponentType<CommonCalculationOptionProps>;
  anchorEl: HTMLElement | null;
};

function Calculation(props: Props): JSX.Element {
  const value = props.value || Options.none.value;
  const valueOption = Options[value];
  const intl = useIntl();

  const option = (
    <props.optionsComponent
      value={value}
      menuOpen={props.menuOpen}
      onClose={props.onMenuClose}
      onChange={props.onChange}
      anchorEl={props.anchorEl}
      property={props.property}
    />
  );

  return (
    // tabindex is needed to make onBlur work on div.
    // See this for more details-
    // https://stackoverflow.com/questions/47308081/onblur-event-is-not-firing
    <div
      className={`Calculation ${value} ${props.class} ${props.menuOpen ? 'menuOpen' : ''} ${
        props.hovered ? 'hovered' : ''
      }`}
      style={props.style}
      onClick={(e) => (props.menuOpen ? props.onMenuClose() : props.onMenuOpen(e.currentTarget))}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex={0}
      // onBlur={props.onMenuClose}
    >
      <div>{option}</div>

      <span className='calculationLabel'>{optionDisplayNameString(valueOption!, intl)}</span>

      {value === Options.none.value && <ExpandMoreIcon fontSize='small' />}

      {value !== Options.none.value && (
        <span className='calculationValue'>
          {Calculations[value] ? Calculations[value](props.cards, props.property, intl) : ''}
        </span>
      )}
    </div>
  );
}

export default Calculation;
