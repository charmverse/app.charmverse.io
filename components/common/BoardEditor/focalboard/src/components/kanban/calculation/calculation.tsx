import { useIntl } from 'react-intl';

import type { IPropertyTemplate } from '../../../blocks/board';
import type { Card } from '../../../blocks/card';
import Button from '../../../widgets/buttons/button';
import Calculations from '../../calculations/calculations';

import { KanbanCalculationOptions } from './calculationOptions';

type Props = {
    cards: Card[];
    cardProperties: IPropertyTemplate[];
    menuOpen: boolean;
    onMenuClose: () => void;
    onMenuOpen: () => void;
    onChange: (data: { calculation: string, propertyId: string }) => void;
    value: string;
    property: IPropertyTemplate;
    readOnly: boolean;
}

function KanbanCalculation (props: Props): JSX.Element {
  const intl = useIntl();

  return (
    <div className='KanbanCalculation'>
      <Button
        onClick={() => (props.menuOpen ? props.onMenuClose : props.onMenuOpen)()}
        onBlur={props.onMenuClose}
        title={Calculations[props.value] ? Calculations[props.value](props.cards, props.property, intl) : ''}
      >
        {Calculations[props.value] ? Calculations[props.value](props.cards, props.property, intl) : ''}
      </Button>

      {
                !props.readOnly && props.menuOpen && (
                  <KanbanCalculationOptions
                    value={props.value}
                    property={props.property}
                    menuOpen={props.menuOpen}
                    onChange={(data: { calculation: string, propertyId: string }) => {
                      props.onChange(data);
                      props.onMenuClose();
                    }}
                    cardProperties={props.cardProperties}
                  />
                )
            }
    </div>
  );
}

export {
  KanbanCalculation
};

