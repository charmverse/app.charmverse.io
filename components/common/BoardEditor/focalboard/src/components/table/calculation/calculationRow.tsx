import React, { useState } from 'react';

import type { Board, IPropertyTemplate } from '../../../blocks/board';
import { createBoard } from '../../../blocks/board';
import type { BoardView } from '../../../blocks/boardView';
import type { Card } from '../../../blocks/card';
import { Constants } from '../../../constants';
import mutator from '../../../mutator';
import Calculation from '../../calculations/calculation';
import { Options } from '../../calculations/options';
import { columnWidth } from '../tableRow';

import { TableCalculationOptions } from './tableCalculationOptions';

type Props = {
    board: Board;
    cards: Card[];
    activeView: BoardView;
    resizingColumn: string;
    offset: number;
    readOnly: boolean;
}

function CalculationRow (props: Props): JSX.Element {

  const [showOptions, setShowOptions] = useState<Map<string, boolean>>(new Map<string, boolean>());

  const toggleOptions = (templateId: string, show: boolean) => {
    const newShowOptions = new Map<string, boolean>(showOptions);
    newShowOptions.set(templateId, show);
    setShowOptions(newShowOptions);
  };

  const titleTemplate: IPropertyTemplate = {
    id: Constants.titleColumnId
  } as IPropertyTemplate;

  const templates: IPropertyTemplate[] = [
    titleTemplate,
    ...props.board.fields.cardProperties.filter((template) => props.activeView.fields.visiblePropertyIds.includes(template.id))
  ];

  const selectedCalculations = props.board.fields.columnCalculations || [];

  const [hovered, setHovered] = useState(false);

  return (
    <div
      className='CalculationRow octo-table-row'
      onMouseEnter={() => setHovered(!props.readOnly)}
      onMouseLeave={() => setHovered(false)}
    >
      {
                templates.map((template) => {
                  const style = { width: columnWidth(props.resizingColumn, props.activeView.fields.columnWidths, props.offset, template.id) };
                  const defaultValue = template.id === Constants.titleColumnId ? Options.count.value : Options.none.value;
                  const value = selectedCalculations[template.id] || defaultValue;

                  return (
                    <Calculation
                      key={template.id}
                      style={style}
                      class={`octo-table-cell ${props.readOnly ? 'disabled' : ''}`}
                      value={value}
                      menuOpen={Boolean(props.readOnly ? false : showOptions.get(template.id))}
                      onMenuClose={() => toggleOptions(template.id, false)}
                      onMenuOpen={() => toggleOptions(template.id, true)}
                      onChange={(v: string) => {
                        const calculations = { ...selectedCalculations };
                        calculations[template.id] = v;
                        const newBoard = createBoard({ block: props.board });
                        newBoard.fields.columnCalculations = calculations;
                        mutator.updateBlock(newBoard, props.board, 'update_calculation');
                        setHovered(false);
                      }}
                      cards={props.cards}
                      hovered={hovered}
                      optionsComponent={TableCalculationOptions}
                      property={template}
                    />
                  );
                })
            }
    </div>
  );
}

export default CalculationRow;
