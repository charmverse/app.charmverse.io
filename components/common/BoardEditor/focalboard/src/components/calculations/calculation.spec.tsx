import { render } from '@testing-library/react';

import { TestBlockFactory } from '../../test/testBlockFactory';
import { wrapIntl } from '../../testUtils';
import { TableCalculationOptions } from '../table/calculation/tableCalculationOptions';

import Calculation from './calculation';

describe('components/calculations/Calculation', () => {
  const board = TestBlockFactory.createBoard();

  const card = TestBlockFactory.createCard(board);
  card.fields.properties.property_2 = 'Foo';
  card.fields.properties.property_3 = 'Bar';
  card.fields.properties.property_4 = 'Baz';

  const card2 = TestBlockFactory.createCard(board);
  card2.fields.properties.property_2 = 'Lorem';
  card2.fields.properties.property_3 = '';
  card2.fields.properties.property_4 = 'Baz';

  test('should match snapshot - none', () => {
    const component = wrapIntl(
      <Calculation
        style={{}}
        class='fooClass'
        value='none'
        menuOpen={false}
        onMenuClose={() => {}}
        onMenuOpen={() => {}}
        onChange={() => {}}
        cards={[card, card2]}
        anchorEl={null}
        hovered={true}
        property={{
          id: 'property_2',
          name: '',
          type: 'text',
          options: []
        }}
        optionsComponent={TableCalculationOptions}
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('should match snapshot - count', () => {
    const component = wrapIntl(
      <Calculation
        style={{}}
        anchorEl={null}
        class='fooClass'
        value='count'
        menuOpen={false}
        onMenuClose={() => {}}
        onMenuOpen={() => {}}
        onChange={() => {}}
        cards={[card, card2]}
        hovered={true}
        property={{
          id: 'property_2',
          name: '',
          type: 'text',
          options: []
        }}
        optionsComponent={TableCalculationOptions}
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('should match snapshot - countValue', () => {
    const component = wrapIntl(
      <Calculation
        style={{}}
        anchorEl={null}
        class='fooClass'
        value='countValue'
        menuOpen={false}
        onMenuClose={() => {}}
        onMenuOpen={() => {}}
        onChange={() => {}}
        cards={[card, card2]}
        hovered={true}
        property={{
          id: 'property_3',
          name: '',
          type: 'text',
          options: []
        }}
        optionsComponent={TableCalculationOptions}
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('should match snapshot - countUniqueValue', () => {
    const component = wrapIntl(
      <Calculation
        style={{}}
        anchorEl={null}
        class='fooClass'
        value='countUniqueValue'
        menuOpen={false}
        onMenuClose={() => {}}
        onMenuOpen={() => {}}
        onChange={() => {}}
        cards={[card, card2]}
        hovered={true}
        property={{
          id: 'property_4',
          name: '',
          type: 'text',
          options: []
        }}
        optionsComponent={TableCalculationOptions}
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
