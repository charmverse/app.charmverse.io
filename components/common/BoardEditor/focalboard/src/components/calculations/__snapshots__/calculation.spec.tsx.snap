// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/calculations/Calculation should match snapshot - count 1`] = `
<div>
  <span
    class="variable variable"
  >
    <div
      class="Calculation count fooClass  hovered"
      tabindex="0"
    >
      <div />
      <span
        class="calculationLabel"
      >
        Count
      </span>
      <span
        class="calculationValue"
      >
        2
      </span>
    </div>
  </span>
</div>
`;

exports[`components/calculations/Calculation should match snapshot - countUniqueValue 1`] = `
<div>
  <span
    class="variable variable"
  >
    <div
      class="Calculation countUniqueValue fooClass  hovered"
      tabindex="0"
    >
      <div />
      <span
        class="calculationLabel"
      >
        Unique
      </span>
      <span
        class="calculationValue"
      >
        1
      </span>
    </div>
  </span>
</div>
`;

exports[`components/calculations/Calculation should match snapshot - countValue 1`] = `
<div>
  <span
    class="variable variable"
  >
    <div
      class="Calculation countValue fooClass  hovered"
      tabindex="0"
    >
      <div />
      <span
        class="calculationLabel"
      >
        Values
      </span>
      <span
        class="calculationValue"
      >
        1
      </span>
    </div>
  </span>
</div>
`;

exports[`components/calculations/Calculation should match snapshot - none 1`] = `
<div>
  <span
    class="variable variable"
  >
    <div
      class="Calculation none fooClass  hovered"
      tabindex="0"
    >
      <div />
      <span
        class="calculationLabel"
      >
        Calculate
      </span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-qolhez-MuiSvgIcon-root"
        data-testid="ExpandLessIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
        />
      </svg>
    </div>
  </span>
</div>
`;
